framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -381,9 +378,80 @@ public void processClassTree(ClassTree classTree) {
 
         checkExtendsImplements(classTree);
 
+        checkPolymorphicClass(classTree);
+
         super.visitClass(classTree, null);
     }
 
+    /**
+     * Throws an error if {@code classTree} has polymorphic fields but is not annotated with
+     * {@code @HasQualifierParameter}.
+     */
+    private void checkPolymorphicClass(ClassTree classTree) {
+        List<? extends Tree> members = classTree.getMembers();
+        Set<? extends AnnotationMirror> topAnnotations =
+                atypeFactory.getQualifierHierarchy().getTopAnnotations();
+
+        for (Tree mem : members) {
+            if (mem.getKind() == Tree.Kind.VARIABLE) {
+                AnnotatedTypeMirror fieldAnno = atypeFactory.getAnnotatedType(mem);
+                for (AnnotationMirror top : topAnnotations) {
+                    AnnotationMirror polyAnnoInHierarchy =
+                            atypeFactory.getQualifierHierarchy().getPolymorphicAnnotation(top);
+                    if (polyAnnoInHierarchy != null) {
+                        PolyTypeScanner polyScanner = new PolyTypeScanner();
+                        if (polyScanner.visit(fieldAnno)) {
+                            Element classTreeElement = TreeUtils.elementFromTree(classTree);
+                            if (!atypeFactory.hasQualifierParameter(classTreeElement)) {
+                                checker.report(
+                                        Result.failure("invalid.polymorphic.qualifier.use"), mem);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /** A scanner that indicates whether any (sub-)types are annotated as polymorphic. */
+    class PolyTypeScanner extends SimpleAnnotatedTypeScanner<Boolean, Void> {
+        @Override
+        protected Boolean reduce(Boolean r1, Boolean r2) {
+            r1 = r1 == null ? false : r1;
+            r2 = r2 == null ? false : r2;
+            return r1 || r2;
+        }
+
+        @Override
+        protected Boolean defaultAction(AnnotatedTypeMirror type, Void aVoid) {
+            if (type == null) {
+                return false;
+            }
+            Set<? extends AnnotationMirror> topAnnotations =
+                    atypeFactory.getQualifierHierarchy().getTopAnnotations();
+            for (AnnotationMirror top : topAnnotations) {
+                AnnotationMirror polyAnnoInHierarchy =
+                        atypeFactory.getQualifierHierarchy().getPolymorphicAnnotation(top);
+                AnnotationMirror annoInHierarchy = type.getAnnotationInHierarchy(top);
+                if (annoInHierarchy != null
+                        && AnnotationUtils.areSameByName(polyAnnoInHierarchy, annoInHierarchy)) {
+                    return true;
+                }
+            }
+            if (type.getKind() == TypeKind.INT
```

Comment:
Why is this if statement needed?  (Also, if it is needed you can replace it with `type.getKind().isPrimitive()`)

Reply:
I'm not sure why but without this check, the test case: https://github.com/typetools/checker-framework/blob/master/framework/tests/all-systems/Issue1003.java
throws a null pointer exception while visiting an int field